src/approximate-qvm.lisp

```diff
@@ -0,0 +1,183 @@
+  "Helper fucntion to check that a key value pair in READOUT-POVMS is valid. The QUBIT must be a NON-NEGATIVE INTEGER, and the POVM must be a correctly formatted POVM."
+  (assert (and (integerp qubit) (<= 0 qubit)))
+  (check-povm povm))
+
+
+(defun (setf qubit-t1) (t1 qvm qubit)
+  "Save the kraus operators for the T1 provided on the specific qubit."
+  (let ((kraus-ops (generate-damping-kraus-map t1 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t1-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the T2  provided on the specified qubit."
+  (let ((kraus-ops (generate-dephasing-kraus-map t2 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t2-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-depolarization) (depol-prob qvm qubit)
+  "Save the kraus operators for the depolarizing channel on QUBIT. The depolarizing channel represents the average gate overrotation/underrotation noise. It is specified by DEPOL-PROB, which is the probability of overrotation/underrotation for the given QUBIT."
+  (let ((kraus-ops (generate-depolarizing-kraus-map depol-prob)))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (depol-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+(defmethod apply-all-kraus-maps ((qvm approx-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Apply all the kraus operators for each type of noise
+  ;; source. KRAUS-OPS is a list of ( (t1-operators) (t2-operators)),
+  ;; where t1-operators is a list of t1 noise maps for each qubit in
+  ;; the instruction.  If the instruction is a 2q gate, we need to
+  ;; tensor the kraus operators for each qubit's error source.
+  (loop :for noise-operators :in kraus-ops
+        :do (let ((simplified-kraus-ops  (reduce #'kraus-kron noise-operators)))
+              (when simplified-kraus-ops
+                (apply-kraus-map qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm approx-qvm) (instr quil:gate-application))
+  ;; For the current instruction in the program, apply the gate
+  ;; corresponding to that instruction, and then apply all the kraus
+  ;; operators for all the sources of noise following that
+  ;; instruction.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q :in qubits :collect (gethash q (t1-ops qvm))))
```

Comment:
What if any of these lists contain `nil`?

If they never should, I think we should add `assert`s.

Reply:
If one of the lists is NIL (ie, the user doesn't specify a T2 for a qubit in `instr`), then in `apply-all-kraus-maps` the NIL is just passed over (so nothing is done to the state, which is what should happen)